main.go

```diff
@@ -0,0 +1,180 @@
+executed directly or through a symlink that doesn't point to an iptables binary,
+it will enter an infinite loop, calling itself recursively.
+
+It's important to note that this proxy behavior will only happen on the first iptables-*
+execution. Following invocations will use directly the binaries for the selected mode.
+*/
+package main
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+)
+
+// iptablesMode represents the two different modes iptables can be
+// configured to: nft or legacy
+type iptablesMode int
+
+const (
+	undefined iptablesMode = iota
+	legacy
+	nft
+
+	iptablesNFTMode    string = "nft"
+	iptablesLegacyMode string = "legacy"
+)
+
+// String returns the iptables mode in string form as it's used in
+// all `iptables-*` commands: `nft` and `legacy`
+func (m iptablesMode) String() string {
+	switch m {
+	case legacy:
+		return iptablesLegacyMode
+	case nft:
+		return iptablesNFTMode
+	default:
+		return "undefined"
+	}
+}
+
+func main() {
+	ctx := context.Background()
+
+	sbinPath, err := detectIptablesDir()
+	if err != nil {
+		logErrorAndExit(err)
+	}
+
+	// We use `xtables-<mode>-multi` binaries by default to inspect the installed rules,
+	// but this can be changed to directly use `iptables-<mode>-save` binaries.
+	iptables := newXtablesMultiInstallation(sbinPath)
+	mode := detectIptablesMode(ctx, iptables)
+
+	// This re-executes the exact same command passed to this program
+	binaryPath := os.Args[0]
+	var args []string
+	if len(os.Args) > 1 {
+		args = make([]string, 0, len(os.Args)-1)
+		args = append(args, os.Args[1:]...)
+	}
+
+	selector := buildIptablesAlternativeSelector(sbinPath)
+	if err := selector.useMode(ctx, mode); err != nil {
+		fmt.Printf("Unable to redirect iptables binaries. (Are you running in an unprivileged pod?): %s\n", err)
+		// fake it, though this will probably also fail if they aren't root
+		binaryPath = xtablesPath(sbinPath, mode)
+		args = os.Args
+	}
+
+	cmdIptables := exec.CommandContext(ctx, binaryPath, args...)
+	cmdIptables.Stdout = os.Stdout
+	cmdIptables.Stderr = os.Stderr
+
+	if err := cmdIptables.Run(); err != nil {
+		code := 1
+		var exitErr *exec.ExitError
+		if errors.As(err, &exitErr) {
+			code = exitErr.ExitCode()
+		} else {
+			// If it's not an ExitError, the command probably didn't finish and something
+			// else failed, which means it might not had outputted anything. In that case,
+			// print the error message just in case.
+			fmt.Fprintf(os.Stderr, "Error: %s\n", err)
+		}
+		os.Exit(code)
+	}
+}
+
+// iptablesInstallation represents the set of iptables-*-save binaries installed in a machine.
+// It is expected the machine supports both nft and legacy modes. This can be implemented by
+// calling directly iptables-*-save, xtables, etc. The implementation should accept the same
+// command arguments as the mentioned binaries.
+type iptablesInstallation interface {
```

Comment:
If you're not including unit tests, then this also seems like unnecessary abstraction...